linker/Linker.Steps/MarkStep.cs

```diff
@@ -217,6 +220,35 @@ void ProcessVirtualMethods ()
 			}
 		}
 
+		void ProcessActivatorCreateInstanceType ()
+		{
+			var markedTypes = Annotations.GetMarkedTypes ();
+			foreach (var information in _activatorCreateInstanceTypes) {
+				_context.Tracer.Push (information.CallingBody.Method);
+				try {
+					foreach (var markedType in markedTypes) {
```

Comment:
Why do you have to process all markedTypes when you already have a possible candidate in CastType?

Reply:
In the cast scenario, the `CastType` is only 1 of N possible types that we need to mark.  We also need to apply the same marking logic to all other marked types derived from `CastType`.  We need to do this because in the cast scenario we don't actually know what type is being created.

See the test `UnusedDerivedTypesAreNotMarked`  https://github.com/mono/linker/pull/419/files#diff-b0f8eca4ab2e651d4eda74da6a827d1b

Reply:
Another approach I could have taken would have been to store information about derived types in `AnnotationsStore` during the `TypeMapStep`.  Then I could have done something like
```
foreach (var derivedType in Annotations.TypesDerivedFrom(information.CastType))
    if (Annotations.IsMark(derivedType))
        ....
```

I didn't take that approach because

* It would be more work
* It would be more complicated.
* Back to my assumption again (which could be a bad assumption to make) that the number of items in `_activatorCreateInstanceTypes` will be relatively small so I went with simple over optimal.